recipe/2024-01-support-private-terraform-repository.md

```diff
@@ -0,0 +1,286 @@
++            namespace: <namespace>
++          }
++       }
++     }
++   }
+    recipes: {      
+      'Applications.Datastores/mongoDatabases':{
+        default: {
+          templateKind: 'terraform'
+          templatePath: 'https://dev.azure.com/test-private-repo'
+        }
+      }
+    }
+  }
+}
+```
+
+Option 3 : Using Applications.Core/secretStores
+
+Use secretStore to store the username and personal access token and add the secret to the new property recipeConfig in the environment.
+
+```diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
++   recipeConfig: {
++     terraform: {
++       secrets: {
++         "dev.azure.com": {
++            secret: secretStore.id
++          }
++       }
++     }
++   }
+    recipes: {      
+      'Applications.Datastores/mongoDatabases':{
+        default: {
+          templateKind: 'terraform'
+          templatePath: 'https://dev.azure.com/test-private-repo'
+        }
+      }
+    }
+  }
+}
+
+resource secretStore 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'github'
+  properties:{
+-  	app: app.id
+    type: 'generic'
+    data: {
+      'pat': {
+        value: '<personal-access-token>'
+      }
+      'username': {
+        value: '<username>'
+      }
+    }
+  }
+}
+
+```
+SecretStore resource also provides an option to use the existing secret, which makes it better way store credentials. But todays secret store implementation is tied to application scope and in this case, secretStore needs to be created before application and environment creation. So we need to change scope of secret store resource to global (by removing the required flag for application property).
+And the referenced secret is retrieved from the secretStore in configLoader and secret details i.e personal-access-token and username as used to modify the template path to this format:
+```
+"git::https://{username}:{PERSONAL_ACCESS_TOKEN}@example.com.com/test-private-repo.git"
+```
+that will be used for recipe deployment.
+
+### API design (if applicable)
+***Model changes***
+
+Addition of new property `recipeConfig` to environment properties.
+
+```diff
+@doc("Environment properties")
+model EnvironmentProperties {
+  @doc("The status of the asynchronous operation.")
+  @visibility("read")
+  provisioningState?: ProvisioningState;
+
+  @doc("The compute resource used by application environment.")
+  compute: EnvironmentCompute;
+
+  @doc("Cloud providers configuration for the environment.")
+  providers?: Providers;
+
+  @doc("Simulated environment.")
+  simulated?: boolean;
+
+  @doc("Specifies Recipes linked to the Environment.")
+  recipes?: RecipeProperties;
+
++  @doc("Specifies recipe configurations needed for the recipes.")
++  recipeConfig?: Record<Record<RecipeConfigProperties>>;
```

Comment:
What's the purpose of the two levels of `Record<>` here? This doesn't seem to match the example.